Fix upgrade_scenario_10

alsrdn opened the pull request:
* Fixes invocation of the `global-state-update-gen` in order to correctly generate the missing system contract registry.
* Fixes an issue where the node tries to cache the system contract registry when initializing the contract runtime component, but since the node was not upgraded yet and the previous version didn't have the system contract registry written the initialization would fail. Now caching the system contract registry on initialization will not yield a fatal error and allows it to be created after the upgrade or when needed.

Fixes: https://github.com/casper-network/casper-node/issues/3621

fizyk20 committed:
Fix the calls to global-state-update-gen in upgrades across 1.4.0 boundary

alsrdn committed:
ContractRuntime: delay init of system contract registry cache

We currently cache the system contract registry in the ContractRuntime
component early in the initialization process although it is not needed
until later. This is a problem when upgrading from a version that did
not have the system contract registry written in the global state.
Allow the ContractRuntime to skip initialization of the system contract
registry until it is actually needed. This allows the upgrade to create
the registry if required and initialize the cache afterwards.

Signed-off-by: Alexandru Sardan <alexandru@casperlabs.io>

Fraser999 commented on node/src/components/contract_runtime.rs:
```diff
@@ -721,6 +726,13 @@ impl ContractRuntime {
             Ok(system_contract_registry) => {
                 self.system_contract_registry = Some(system_contract_registry);
             }
+            Err(EngineStateError::MissingSystemContractRegistry) => {
+                // The system contract registry might be missing because the node did not have it
+                // before and may create it through an upgrade. Since this is a
+                // cache, allow it to be populated later.
+                self.system_contract_registry = None;
+                warn!("ContractRuntime: system contract registry cache was not initialized.");
+            }
```
I wonder if we can have this be a hard error (i.e. log an `error!` and return `Err`) if we end up here via the `MainReactor::refresh_contract_runtime` flow.  For cases where we end up here via `MainReactor::commit_(genesis|upgrade)` it can be ignored (i.e. warn or even info only).

I also don't think we need to set `self.system_contract_registry = None;` here do we?  It would already be `None` I assume, but if it wasn't we shouldn't be overwriting it.  Maybe worth a `debug_assert!` that it's `None`?

alsrdn replied:
I don't think we should distinguish here if we should log an error here or not depending on the call site. In essence the `self.system_contract_registry` is an internal cache of the contract runtime and it needs to be managed internally by the component. What we're saying here is that we initialized the contract runtime, it is usable, but we couldn't yet cache the system contract registry. It shouldn't fail if it's not needed at this point and should fail when it is used, if that is truly critical, which I don't think it even is right now.

I would apply the same argument for setting it to `None` here, though this case is a little weird since it means that something went wrong with the global state since the system contract registry is stable and should always be present if it has been in the past. Maybe we should log a different warning in this case stating what happened. WDYT?

alsrdn replied:
As discussed, I left only the lazy initialization of the cache at use time.

alsrdn committed:
contract_runtime: don't init system contract registry cache when setting
initial state

Only initialize the system contract registry cache in the contract
runtime when it is first used.
Use the state root hash from the execution pre-state to try and
initialize the cache since the registry is stable in the chain.

Signed-off-by: Alexandru Sardan <alexandru@casperlabs.io>

rafal-ch commented on node/src/components/contract_runtime.rs:
```diff
@@ -375,6 +375,9 @@ impl ContractRuntime {
                 let engine_state = Arc::clone(&self.engine_state);
                 let metrics = Arc::clone(&self.metrics);
+
+                self.try_init_system_contract_registry_cache();
```
A nit, but since this function is infallible, shoud it be just `init_system_contract_registry_cache()`?

alsrdn replied:
Indeed it does not fail but it also does not guarantee that the system contract registry cache will be set either. I don't have a strong preference for naming it. However I think that if the code changes or something goes wrong a reader might be more inclined to investigate that function if it has a `try_` prefix.

rafal-ch commented on node/src/components/contract_runtime.rs:
```diff
@@ -921,6 +899,25 @@ impl ContractRuntime {
         &self.engine_state
     }
+
+    #[inline]
```
Why we would like this to be inlined?

alsrdn replied:
This code is small enough to go into `handle_contract_runtime_request` actually though I preferred to move it for better readability. It doesn't have to be inline, I just considered to mark it that way since it has a single call site and is small enough to be a good candidate for it.

rafal-ch replied:
Sure, thanks for explanation :+1:

rafal-ch reviewed (approved):
LGTM :+1: 
With minor nits.

alsrdn commented:
bors merge
